File: nuTerra/TextureEngine/Quadtree.vb
Public Class Quadtree
	Public Level As Integer
	Public Rectangle As Rectangle
	Public Mapping As Point

	Public Children As Quadtree()

	Public Sub New(rect As Rectangle, level As Integer)
		Me.Level = level
		Me.Rectangle = rect
		ReDim Me.Children(3)
	End Sub

	Public Sub Add(request As Page, mapping As Point)
		Dim scale = 1 << request.Mip
		Dim x = request.X * scale
		Dim y = request.Y * scale

		Dim node = Me
		While request.Mip < node.Level
			For i = 0 To 3
				Dim rect = GetRectangle(node, i)
				If rect.Contains(x, y) Then
					' Create a New one if needed
					If node.Children(i) Is Nothing Then
						node.Children(i) = New Quadtree(rect, node.Level - 1)
						node = node.Children(i)
						Exit For
					Else ' Otherwise traverse the tree
						node = node.Children(i)
						Exit For
					End If
				End If
			Next
		End While
		' We have created the correct node, now set the mapping
		node.Mapping = mapping
	End Sub

	Public Sub Remove(request As Page)
		Dim index As Integer
		Dim node = FindPage(Me, request, index)
		If node IsNot Nothing Then
			node.Children(index) = Nothing
		End If
	End Sub

	Public Sub Write(data As UShort(,), miplevel As Integer)
		Quadtree.Write(Me, data, miplevel)
	End Sub

	' Static Functions
	Shared Function GetRectangle(node As Quadtree, index As Integer) As Rectangle
		Dim x = node.Rectangle.X
		Dim y = node.Rectangle.Y
		Dim w = node.Rectangle.Width \ 2
		Dim h = node.Rectangle.Height \ 2

		Select Case index
			Case 0
				Return New Rectangle(x, y, w, h)
			Case 1
				Return New Rectangle(x + w, y, w, h)
			Case 2
				Return New Rectangle(x + w, y + h, w, h)
			Case 3
				Return New Rectangle(x, y + h, w, h)
		End Select

		Throw New ArgumentOutOfRangeException("index")
	End Function

	Shared Sub Write(node As Quadtree, data As UShort(,), miplevel As Integer)
		If node.Level >= miplevel Then
			Dim rx = node.Rectangle.X >> miplevel
			Dim ry = node.Rectangle.Y >> miplevel
			Dim rw = node.Rectangle.Width >> miplevel
			Dim rh = node.Rectangle.Height >> miplevel

			For i = ry To ry + rh - 1
				For j = rx To rx + rw - 1
					If node.Mapping.X > 32 Then Stop
					If node.Mapping.Y > 64 Then Stop
					If node.Level > 32 Then Stop
					data(i, j) = (node.Mapping.X << 11) Or (node.Mapping.Y << 5) Or node.Level
				Next
			Next

			For Each child In node.Children
				If child IsNot Nothing Then
					Quadtree.Write(child, data, miplevel)
				End If
			Next
		End If
	End Sub

	Shared Function FindPage(node As Quadtree, request As Page, ByRef index As Integer) As Quadtree
		Dim scale = 1 << request.Mip
		Dim x = request.X * scale
		Dim y = request.Y * scale

		' Find the parent of the child we want to remove
		Dim exitloop = False
		While Not exitloop
			exitloop = True

			For i = 0 To 3
				If node.Children(i) IsNot Nothing AndAlso node.Children(i).Rectangle.Contains(x, y) Then
					' We found it
					If request.Mip = node.Level - 1 Then
						index = i
						Return node
					Else ' Check the children
						node = node.Children(i)
						exitloop = False
					End If
				End If
			Next
		End While

		' We couldn't find it so it must not exist anymore
		index = -1
		Return Nothing
	End Function
End Class

File: nuTerraCPP/Quadtree.cpp
#include <algorithm>
#include "Quadtree.h"

Quadtree::Quadtree(Rect _rect, int _level)
	: m_rectangle(_rect)
	, m_level(_level)
{
	for (int i = 0; i < 4; ++i)
	{
		m_children[i] = nullptr;
	}
}

Quadtree::~Quadtree()
{
	for (int i = 0; i < 4; ++i)
	{
		if (m_children[i] != nullptr)
		{
			delete m_children[i];
		}
	}
}

void Quadtree::add(uint32_t request, int mapping)
{
	const int reqMip = request & 0xFF;
	const int reqY = (request >> 8) & 0xFFF;
	const int reqX = (request >> 20) & 0xFFF;

	int scale = 1 << reqMip; // Same as pow( 2, mip )
	int x = reqX * scale;
	int y = reqY * scale;

	Quadtree* node = this;

	while (reqMip < node->m_level)
	{
		for (int i = 0; i < 4; ++i)
		{
			auto rect = node->getRectangle(i);
			if (rect.contains(x, y))
			{
				// Create a new one if needed
				if (node->m_children[i] == nullptr)
				{
					node->m_children[i] = new Quadtree(rect, node->m_level - 1);
					node = node->m_children[i];
					break;
				}
				// Otherwise traverse the tree
				else
				{
					node = node->m_children[i];
					break;
				}
			}
		}
	}

	// We have created the correct node, now set the mapping
	node->m_mapping = mapping;
}

void Quadtree::remove(uint32_t request)
{
	int  index;
	auto node = findPage(this, request, &index);

	if (node != nullptr)
	{
		delete node->m_children[index];
		node->m_children[index] = nullptr;
	}
}

void Quadtree::write(uint16_t* data, int size, int miplevel)
{
	write(this, data, size, miplevel);
}

Rect Quadtree::getRectangle(int index)
{
	int x = m_rectangle.m_x;
	int y = m_rectangle.m_y;
	int w = m_rectangle.m_width / 2;
	int h = m_rectangle.m_width / 2;

	switch (index)
	{
	case 0: return { x    , y    , w, h };
	case 1: return { x + w, y    , w, h };
	case 2: return { x + w, y + h, w, h };
	case 3: return { x    , y + h, w, h };
	default: break;
	}

	return { 0, 0, 0, 0 };
}

void Quadtree::write(Quadtree* node, uint16_t* data, const int size, const int miplevel)
{
	if (node->m_level >= miplevel)
	{
		const int rx = node->m_rectangle.m_x >> miplevel;
		const int ry = node->m_rectangle.m_y >> miplevel;
		const int rw = (node->m_rectangle.m_width >> miplevel);
		const int rh = ry + (node->m_rectangle.m_width >> miplevel);

		const uint16_t value = (node->m_mapping << 5) | node->m_level;
		for (int i = ry; i < rh; ++i)
		{
			uint16_t* ptr = data + i * size + rx;
			const uint16_t* end = ptr + rw;
			for (; ptr < end; ++ptr)
			{
				*ptr = value;
			}
		}

		for (int i = 0; i < 4; ++i)
		{
			auto child = node->m_children[i];
			if (child != nullptr)
			{
				Quadtree::write(child, data, size, miplevel);
			}
		}
	}
}

Quadtree* Quadtree::findPage(Quadtree* node, uint32_t request, int* index)
{
	const int reqMip = request & 0xFF;
	const int reqY = (request >> 8) & 0xFFF;
	const int reqX = (request >> 20) & 0xFFF;

	int scale = 1 << reqMip; // Same as pow( 2, mip )
	int x = reqX * scale;
	int y = reqY * scale;

	// Find the parent of the child we want to remove
	bool exitloop = false;
	while (!exitloop)
	{
		exitloop = true;
		for (int i = 0; i < 4; ++i)
		{
			if (node->m_children[i] != nullptr && node->m_children[i]->m_rectangle.contains(x, y))
			{
				// We found it
				if (reqMip == node->m_level - 1)
				{
					*index = i;
					return node;
				}
				// Check the children
				else
				{
					node = node->m_children[i];
					exitloop = false;
				}
			}
		}
	}

	// We couldn't find it so it must not exist anymore
	*index = -1;
	return nullptr;
}
